회원가입 화면 step4 jake

jryoun1 opened the pull request:
step4까지 구현 완료했습니다.
코드리뷰 부탁드려요~

[구현사항]
1. 요구서에 나와있는 기능들 전부 구현하였습니다.

[step3에서 수정사항]
1. AspectFit -> AspectFill로 다시 수정 (리뷰어 분께서 착각하셔서 변경하였었는데 다시 원래대로 수정)
2. UserInfoTemporarySave 타입의 객체명 변경 checkFilled->userInfoTemporarySave 로 변경

[고민한 점]
1. 두 번째 view인 SignUpViewController에서 세번째 view인 SignUpOptionViewController로의 userInfoTemporarySave 객체를 옮겨주는 방법에 대해서 고민했었는데, 프로퍼티를 사용해서 객체를 옮기는 방법으로 해결하였습니다.

jryoun1 committed:
feat: placeholder 설정, 생년월일display UILabel 추가, text field type phonePad로 설정

jryoun1 committed:
refactor: UserInfoTemporarySave 인스턴스 변수명 checkFilled->userInfoTemporarySave로 변경

jryoun1 committed:
feat: SignUpOptionViewController와 SignUpVeiwController간의 프로퍼티를 이용해 userInfoTemporarySave 객체 보내주기

jryoun1 committed:
feat: SignUpOptionViewController에서의 UIOutlet 프로퍼티 생성, SignUpViewController의 객체를 받을 userInfoTemporarySave 프로퍼티 생성

jryoun1 committed:
feat: dateFormatter와 화면이 전환될때마다 사용자의 phone, date 정보를 띄워주도록 구현

jryoun1 committed:
feat: 사용자로부터 date가 선택되면 값을 UILabel에 띄워주고 userInfoTemporarySave의 birthDate에 저장

jryoun1 committed:
feat: UIAction, UIOutlet 연결

jryoun1 committed:
step3, step4 conflict 해결

jryoun1 committed:
refactor: view간에 프로퍼티를 이용해서 객체 전달 코드 수정

jryoun1 committed:
refactor: displayUserInfo() 메서드에서 옵셔널 바인딩으로 phone 정보 가져오는 걸로 수정

jryoun1 committed:
feat: keyboard Accessory view에 done버튼 추가 구현 및 viewDidLoad()에서 setKeyboardDoneButton() 호출

jryoun1 committed:
feat: phoneTextField에 입력이 끝나면 값을 userInfoTemporarySave의 phone에 저장하도록 구현

jryoun1 committed:
refactor: 메소드 명 goBack()->tapGoBackButton()으로 변경, @IBAction->@objc 로 변경

jryoun1 committed:
refactor: 메소드 접근제어 internal->private으로 변경, text field delegate 선언부분 viewWillAppear()->viewDidLoad()로 변경

jryoun1 committed:
feat: tapCancelButton() 메서드 구현

jryoun1 committed:
feat: tapSignUpButton() 메서드 구현

이때 userInfoTemporarySave 객체의 정보들을 shared에 다 옮겨주고 userInfoTemporarySave 객체는 초기화 시켜준다.
그리고 main 화면으로 돌아가기 실행

jryoun1 committed:
feat: signUpButtonEnableChange() 메서드 구현 및 호출

해당 메서드는 정보가 채워졌는지 확인하고 다 채워졌다면 isEnable을 true로 변경시켜준다

jryoun1 committed:
feat: viewWillAppear()에서 만약 shared에 id가 저장되어있으면 해당 id를 idTextField에 입력

jryoun1 committed:
feat: IBOulet, IBAction 연결

jryoun1 committed:
fix: ImageView에서 다시 AspectFit->AspectFill로 수정

AppleCEO commented on SignUpFlow/SignUpFlow/SignUpOptionViewController.swift:
```diff
@@ -8,8 +8,100 @@
 import UIKit
 
 class SignUpOptionViewController: UIViewController {
```
SignUpOptionViewController 라는 뷰컨트롤러 이름은 어떤 의미인가요?

jryoun1 replied:
@AppleCEO 
처음에는 SignUpViewController의 다음 ViewController라서 SignUpSecondViewController이나 SignUpLastViewController라는 이름을 고민했었습니다. 그러나 이름에 이렇게 순서가 들어간 네이밍은 별로 좋지 않다고 생각했습니다. 
그래서 다른 애플리케이션에서의 회원가입을 생각해보면 이메일, 비밀번호는 필수지만 휴대폰번호나 생년월일은 사용자의 선택, 즉 Option일 수 있다고 생각해서 SignUpOptionViewController라고 네이밍하게되었습니다!!

AppleCEO commented on SignUpFlow/SignUpFlow/SignUpOptionViewController.swift:
```diff
@@ -8,8 +8,100 @@
 class SignUpOptionViewController: UIViewController {
+    @IBOutlet weak var phoneTextField: UITextField!
+    @IBOutlet weak var birthDateDisplayLabel: UILabel!
+    @IBOutlet weak var datePicker: UIDatePicker!
```
데이트 피커를 처음 활용하실텐데 잘 만드셨네요~
데이트 피커의 인라인 스타일은 iOS 14부터 사용할 수 있습니다.
iOS13 이전 버전에서 이 부분은 어떤 식으로 만들 수 있을까요?

AppleCEO commented on SignUpFlow/SignUpFlow/SignUpViewController.swift:
```diff
@@ -14,7 +14,7 @@ class SignUpViewController: UIViewController, UINavigationControllerDelegate {
     @IBOutlet weak var introductionTextView: UITextView!
     @IBOutlet weak var goNextButton: UIButton!
-    let checkFilled = UserInfoTemporarySave()
+    let userInfoTemporarySave = UserInfoTemporarySave()
```
조금 더 이름이 좋아졌네요!
의미가 명확하게 개선해볼 수 있을까요?

AppleCEO commented on SignUpFlow/SignUpFlow/SignUpViewController.swift:
```diff
@@ -80,15 +80,15 @@ class SignUpViewController: UIViewController, UINavigationControllerDelegate {
     
     @IBAction private func tapGoNextButton() {
-        guard let signUpOptionViewController = self.storyboard?.instantiateViewController(withIdentifier: "SignUpOption") else {
+        guard let signUpOptionViewController = self.storyboard?.instantiateViewController(withIdentifier: "SignUpOption") as? SignUpOptionViewController else {
```
타입 캐스팅도 해보셨군요! 고급집니다~

AppleCEO reviewed (commented):
빠르게 구현하셨네요~
데이트 피커에서 날짜를 선택하고 빈 공간을 탭해서 나왔을 때 가입 버튼은 활성화되지 않았습니다.
이전으로 돌아갔다가 다시 SignUpOptionViewController로 돌아오면 가입 버튼이 활성화되었어요.
날짜를 선택 했을 때 가입 버튼이 활성화되는 부분 다시 확인 부탁드릴께요!

jryoun1 committed:
fix: 다음, 가입 버튼의 활성화 조건 검사 시점 변경 textFieldEndEditing->textFieldDidChange

기존에는 사용자가 입력을 마치고 done 버튼이나 화면을 터치해서 endEditing이후에 버튼 활성화 조건을 검사했다면, 변경한 코드는 실시간으로 값이 변경되는 것을 확인하여
버튼의 활성화와 비활성화를 결정한다. 이와같이 수정한 이유는 사용자의 입장에서는 매 시점에서 버튼이 활성화/비활성화 되는 것이 좀 더 편리할 것이라고 생각했기 때문이다.

jryoun1 committed:
refactor: userInfoTemporarySave -> userInfo로 instance 변수명 변경

jryoun1 committed:
refactor: UserInfoTemporarySave 타입의 resetInfo()-> reset()으로 메소드명 변경

yagom commented:
@jryoun1  추가 커밋은 추가 리뷰 요청한건가요? :) 따로 코멘트가 없어서요

jryoun1 commented:
@yagom @AppleCEO 
아직 iOS 버전에 따른 datePicker에 대해서 고민을 하고 있어서 코멘트를 날리지 않았습니다!! 
좀만 더 고민해보고 코멘트 달게요:)

jryoun1 committed:
feat: datePicker iOS 버전에 따라서 style 설정 (14이상 = inline), (13.4이상 = compact)

jryoun1 commented:
@AppleCEO 
14이상의 버전에서는 datePicker의 스타일을 .inline으로 설정해주고, 13.4이상에서는 .compact style로 설정하는 코드를 viewDidLoad()에 넣어주었는데 이러한 방식을 말씀하신 건지 아니면 낮은 버전에서도 .inline을 코드로 그리는 것을 말씀하시는 건지 잘 모르겠습니다ㅠㅠ

AppleCEO commented:
고생하셨습니다. 
다음 단계로 갑시다!
